File: MobileApp/AlexaDeviceFinder/AlexaDeviceFinder.Android/Utilities/OtpEditText.cs
using Android.Content;
using Android.Content.Res;
using Android.Graphics;
using Android.Util;
using Android.Views;
using Android.Views.InputMethods;
using AndroidX.AppCompat.Widget;
using AndroidX.Core.Content;

namespace DeviceFinder.Droid.Utilities
{
    public class OtpEditText : AppCompatEditText
    {
        private ColorStateList defaultTextColors;
        private Context _context;

        public OtpEditText(Context context, IAttributeSet attrs, int defStyle) : base(context, attrs, defStyle)
        {
            defaultTextColors = GetTextColors(context, null);
        }

        public OtpEditText(Context context, IAttributeSet attrs) : base(context, attrs)
        {
            defaultTextColors = GetTextColors(context, null);
        }

        public OtpEditText(Context context) : base(context)
        {
            defaultTextColors = GetTextColors(context, null);
        }

        public void SetErrorState()
        {
            Color errorColor = new Color(ContextCompat.GetColor(Context, Resource.Color.errorred));
            
            Background.Mutate().SetColorFilter(errorColor, PorterDuff.Mode.SrcAtop);
            SetTextColor(errorColor);
        }

        public void ClearErrorState()
        {
            Background.Mutate().ClearColorFilter();
            SetTextColor(defaultTextColors);
        }

        protected void OnFocusChanged(bool hasFocus, FocusSearchDirection direction, Rect previouslyFocusedRect)
        {
            base.OnFocusChanged(hasFocus, direction, previouslyFocusedRect);

            if (hasFocus)
            {
                RequestFocus();
                Text = "";

                ViewGroup viewGroup = Parent as ViewGroup;

                for (int i = 0; i < viewGroup.ChildCount; i++)
                {
                    View child = viewGroup.GetChildAt(i);
                    if (child is OtpEditText)
                        ((OtpEditText)child).ClearErrorState();
                }
            }
        }

        public override IInputConnection OnCreateInputConnection(EditorInfo outAttrs)
        {
            return new OtpInputConnection(base.OnCreateInputConnection(outAttrs), true, this);
        }

        private class OtpInputConnection : InputConnectionWrapper
        {
            private OtpEditText _parentEditText;

            public OtpInputConnection(IInputConnection target, bool mutable, OtpEditText editText) : base(target, mutable)
            {
                this._parentEditText = editText;
            }

            public override bool SendKeyEvent(KeyEvent keyEvent)
            {
                bool keyEventResult = base.SendKeyEvent(keyEvent);

                if (!_parentEditText.HasFocus)
                    return keyEventResult;

                if (keyEvent.Action == KeyEventActions.Down && keyEvent.KeyCode == Keycode.Del)
                {
                    View nextLeftFocus = _parentEditText.FocusSearch(FocusSearchDirection.Left);
                    _parentEditText.Text = ("");

                    if (nextLeftFocus != null)
                        nextLeftFocus.RequestFocus();
                }
                else if (keyEvent.Action == KeyEventActions.Up)
                {
                    if (!char.IsDigit(keyEvent.Number))
                        return false;

                    View nextRightFocus = _parentEditText.FocusSearch(FocusSearchDirection.Right);
                    _parentEditText.Text = (keyEvent.Number + "");

                    if (_parentEditText.Length() > 0 && nextRightFocus != null)
                    {
                        nextRightFocus.RequestFocus();
                    }
                    else if (nextRightFocus == null)
                    {
                        InputMethodManager imm = (InputMethodManager) _parentEditText.Context.GetSystemService(Context.InputMethodService);
                        imm.HideSoftInputFromWindow(_parentEditText.WindowToken, 0);
                        _parentEditText.ClearFocus();
                    }
                }

                return keyEventResult;
            }

            public bool deleteSurroundingText(int beforeLength, int afterLength)
            {
                if (beforeLength == 1 && afterLength == 0)
                {
                    return SendKeyEvent(new KeyEvent(KeyEventActions.Down, Keycode.Del))
                            && SendKeyEvent(new KeyEvent(KeyEventActions.Up, Keycode.Del));
                }
                else
                {
                    return base.DeleteSurroundingText(beforeLength, afterLength);
                }
            }
        }
    }
}

File: AndroidApp/app/src/main/java/com/jamespfluger/devicefinder/controls/OtpEditText.java
package com.jamespfluger.devicefinder.controls;

import static android.content.Context.INPUT_METHOD_SERVICE;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputConnection;
import android.view.inputmethod.InputConnectionWrapper;
import android.view.inputmethod.InputMethodManager;

import androidx.appcompat.widget.AppCompatEditText;

import com.jamespfluger.devicefinder.R;

import java.util.Locale;

public class OtpEditText extends AppCompatEditText {
    private final ColorStateList defaultTextColors = getTextColors();

    public OtpEditText(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public OtpEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public OtpEditText(Context context) {
        super(context);
    }

    public void setErrorState() {
        getBackground().mutate().setColorFilter(getResources().getColor(R.color.red_error), PorterDuff.Mode.SRC_ATOP);
        setTextColor(getResources().getColor(R.color.red_error));
    }

    public void clearErrorState() {
        getBackground().mutate().clearColorFilter();
        setTextColor(defaultTextColors);
    }

    @Override
    protected void onFocusChanged(boolean hasFocus, int direction, Rect previouslyFocusedRect) {
        super.onFocusChanged(hasFocus, direction, previouslyFocusedRect);

        if (hasFocus) {
            requestFocus();
            setText("");

            ViewGroup viewGroup = (ViewGroup) getParent();

            for (int i = 0; i < viewGroup.getChildCount(); i++) {
                View child = viewGroup.getChildAt(i);
                if (child instanceof OtpEditText) {
                    ((OtpEditText) child).clearErrorState();
                }
            }
        }
    }

    @Override
    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
        return new OtpInputConnection(super.onCreateInputConnection(outAttrs), true);
    }

    private class OtpInputConnection extends InputConnectionWrapper {
        public OtpInputConnection(InputConnection target, boolean mutable) {
            super(target, mutable);
        }

        @Override
        public boolean sendKeyEvent(KeyEvent event) {
            boolean keyEventResult = super.sendKeyEvent(event);

            if (!hasFocus()) {
                return keyEventResult;
            }

            if (event.getAction() == KeyEvent.ACTION_DOWN && event.getKeyCode() == KeyEvent.KEYCODE_DEL) {
                View nextLeftFocus = focusSearch(FOCUS_LEFT);
                setText("");

                if (nextLeftFocus != null) {
                    nextLeftFocus.requestFocus();
                }
            } else if (event.getAction() == KeyEvent.ACTION_UP) {
                if (!Character.isDigit(event.getNumber())) {
                    return false;
                }

                View nextRightFocus = focusSearch(FOCUS_RIGHT);
                setText(String.format(Locale.ENGLISH, "%c", event.getNumber()));

                if (getText() != null && getText().length() > 0 && nextRightFocus != null) {
                    nextRightFocus.requestFocus();
                } else if (nextRightFocus == null) {
                    InputMethodManager imm = (InputMethodManager) getContext().getApplicationContext().getSystemService(INPUT_METHOD_SERVICE);
                    imm.hideSoftInputFromWindow(getWindowToken(), 0);
                    clearFocus();
                }
            }

            return keyEventResult;
        }

        @Override
        public boolean deleteSurroundingText(int beforeLength, int afterLength) {
            if (beforeLength == 1 && afterLength == 0) {
                return sendKeyEvent(new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DEL))
                        && sendKeyEvent(new KeyEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_DEL));
            } else {
                return super.deleteSurroundingText(beforeLength, afterLength);
            }
        }
    }
}
